### PR TITLE
Add conditional if else tests

### PR DESCRIPTION
**Pull request by lpawelcz**
This PR needs #177 merged before.

### CODE REVIEW

File: frontends/uhdm/UhdmAst.cc [hunk 1]
```diff
@@ -85,16 +85,29 @@ void UhdmAst::visit_default_expr(vpiHandle obj_h)  {
...
 			if (child->type == AST::AST_INITIAL) {
 				initial_node = child;
+			}
+
+			if (initial_node != nullptr && (*always_child)->type == AST::AST_ALWAYS)
 				break;
+
+			if ((*always_child)->type != AST::AST_ALWAYS) {
+				always_child++;
 			}
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]:
```suggestion
		AST::AstNode* always_child = nullptr;
		for (auto child : mod->children) {
			if (child->type == AST::AST_INITIAL) {
				initial_node = child;
				break;
			} else if (child->type == AST::AST_ALWAYS) {
			        always_child = child;
			        break;
			}
```

We want to add initial node before any another always node, so if we found, that there is already initial node in module, we should stop looking for always and only when we won't find initial node, we need location of always node.

File: frontends/uhdm/UhdmAst.cc [hunk 2]
```diff
@@ -85,16 +85,29 @@ void UhdmAst::visit_default_expr(vpiHandle obj_h)  {
...
+		// Ensure single AST_INITIAL node is located in AST_MODULE
+		// before any AST_ALWAYS
 		if (initial_node == nullptr) {
 			initial_node = new AST::AstNode(AST::AST_INITIAL);
-			mod->children.push_back(initial_node);
+			if ((*always_child)->type == AST::AST_ALWAYS) {
+				mod->children.insert(always_child, 1, initial_node);
+			} else {
+				mod->children.push_back(initial_node);
+			}
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]:
```suggestion
			if (always_child) {
				mod->children.insert(always_child, 1, initial_node);
			} else {
				mod->children.push_back(initial_node);
			}
```

File: frontends/uhdm/UhdmAst.cc [hunk 3]
```diff
@@ -555,7 +568,33 @@ void UhdmAst::process_module() {
...
+										for (;module_child != current_node->children.end(); module_child++) {
+											if ((*module_child)->type == AST::AST_ALWAYS) {
+												break;
+											}
+										}
+										if (module_child != current_node->children.end())
+										// Place AST_INITIAL before AST_ALWAYS
+											current_node->children.insert(module_child, 1, node);
+										else
+											add_or_replace_child(current_node, node);
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Isn't this already done in ``visit_default_expr``? If we already added node, we should return ``nullptr`` and here we will skip adding this node.

File: frontends/uhdm/UhdmAst.cc [hunk 4]
```diff
@@ -555,7 +568,33 @@ void UhdmAst::process_module() {
...
+											if ((*module_child)->type == AST::AST_ALWAYS) {
+												break;
+											}
+										}
+										if (module_child != current_node->children.end())
+										// Place AST_ASSIGN before AST_ALWAYS
+											current_node->children.insert(module_child, 1, node);
+										else
+											add_or_replace_child(current_node, node);
+									} else {
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Why do we need this change? I think this shouldn't be needed.

**Review by kamilrakoczy — CHANGES_REQUESTED** [review]
Generally looks OK, but some changes are required.

File: frontends/uhdm/UhdmAst.cc (hunk 1 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: I modified the code to stop looking for AST_ALWAYS after it has found AST_INITIAL. I did it in a different way because `always_child` must be an iterator that can be used in `insert()`

File: frontends/uhdm/UhdmAst.cc (hunk 2 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: `insert()` requires first argument to be an iterator

File: frontends/uhdm/UhdmAst.cc (hunk 3 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: We need those modifications in `process_module` because this is the place where we can control the node ordering in AST. When this peace of code is deleted, there would be `AST_ALWAYS` with whole *if* statement at the begining of  `AST_MODULE` and after that there would be `AST_ASSIGN` with `a` and `AST_INITIAL` with `b`. Such situation is messing up the whole AST which results in bad behavior of if statements in Yosys.

File: frontends/uhdm/UhdmAst.cc (hunk 4 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Right, this is not needed. It is now removed.

File: frontends/uhdm/UhdmAst.cc (hunk 1 above)
Comment (kamilrakoczy) [reply frontends/uhdm/UhdmAst.cc]: Ok, so I think it would be better to use ``find_if`` in ``if`` below, when there isn't AST_INITIAL node to get the iterator.
Current code is very hard to follow (at least to me) and even variable is named ``always_child`` it is not guaranteed, that behind this iterator there is ``AST_ALWAYS`` node, so it is very confusing.

Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Simplified with `find_if`. Now we just look for `AST_INITIAL` in the loop. Then if it was not found we get the iterator for first `AST_ALWAYS` or for the last child of `AST_MODULE` if AST_ALWAYS was not found. Then we insert the AST_INTIAL at the place of the iterator.

File: frontends/uhdm/UhdmAst.cc [hunk 5]
```diff
@@ -555,7 +561,21 @@ void UhdmAst::process_module() {
 							  obj_h,
 							  [&](AST::AstNode* node) {
 								  if (node) {
+									if (node->type == AST::AST_INITIAL) {
+										auto module_child = current_node->children.begin();
+										for (;module_child != current_node->children.end(); module_child++) {
+											if ((*module_child)->type == AST::AST_ALWAYS) {
+												break;
+											}
+										}
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Please use ``find_if`` here also.

**Review by kamilrakoczy — CHANGES_REQUESTED** [review]
Just 1 small comment, but I see, that other test are failing for some reason. Please check and fix them, beside of that, this change looks fine now.

**Comment by lpawelcz** [comment]
Fixed failing tests, rebased onto newest `uhdm-yosys` and squashed commits into 3 cleaner ones. PR ready for final review.

File: frontends/uhdm/UhdmAst.cc [hunk 6]
```diff
@@ -555,7 +561,14 @@ void UhdmAst::process_module() {
 							  obj_h,
 							  [&](AST::AstNode* node) {
 								  if (node) {
+									if (node->type == AST::AST_INITIAL) {
+										auto insert_it = find_if(current_node->children.begin(), current_node->children.end(),
+												[] (AST::AstNode *node) {return (node->type == AST::AST_ALWAYS);} );
+										// Place AST_INITIAL before AST_ALWAYS if found
+										current_node->children.insert(insert_it, 1, node);
+									} else {
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Correct me, if I'm wrong, but I think this can made design with 2 ``AST_INITIAL`` nodes in the design (and we want only one).

If yes, this part should be moved to ``add_or_replace_child`` and when we are replacing child, we should remove old one and place new one before ``AST_ALWAYS``

Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Yes, You are right. This leads to multiple `AST_INITIAL` nodes in the design. I moved this part to `add_or_replace_child` as a special case for AST_INITIALs. In fact, this piece of code had to be rewritten because we are operating on already existing nodes and we have to move the contents of one node to another ensuring that we have only one AST_INITIAL in the design and  one AST_BLOCK per Initial node. In order to do that a number of specific cases must be handled.

File: frontends/uhdm/UhdmAst.cc [hunk 7]
```diff
@@ -248,8 +248,55 @@ static void add_or_replace_child(AST::AstNode* parent, AST::AstNode* child) {
...
+	} else if (child->type == AST::AST_INITIAL) {
+		// Special case for initials
+		// Ensure that there is only one AST_INITIAL in the design
+		// And there is only one AST_BLOCK inside that initial
+		// Copy nodes from child initial to parent initial
+		auto initial_node_it = find_if(parent->children.begin(), parent->children.end(),
+			[] (AST::AstNode *node) {return (node->type == AST::AST_INITIAL);} );
+		if (initial_node_it != parent->children.end()) {
+			AST::AstNode* initial_node = *initial_node_it;
+			auto block_node_it = find_if(initial_node->children.begin(), initial_node->children.end(),
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Isn't ``AST_BLOCK`` need to be 1st children of ``AST_INITIAL`` (and only children node of ``AST_INITIAL``)? 

If yes, we should just check, if there isn't any node in AST_INITIAL or there is and first children is ``AST_BLOCK`` (and fail if not).

File: frontends/uhdm/UhdmAst.cc [hunk 8]
```diff
@@ -248,8 +248,55 @@ static void add_or_replace_child(AST::AstNode* parent, AST::AstNode* child) {
...
+				[] (AST::AstNode *node) {return (node->type == AST::AST_BLOCK);} );
+			if (block_node_it != initial_node->children.end()) {
+				AST::AstNode* block_node = *block_node_it;
+				auto child_block_node_it = find_if(child->children.begin(), child->children.end(),
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Same as above applies here.

File: frontends/uhdm/UhdmAst.cc [hunk 9]
```diff
@@ -248,8 +248,55 @@ static void add_or_replace_child(AST::AstNode* parent, AST::AstNode* child) {
...
+					// AST_BLOCK in child does not exist, place the contents
+					// of child initial node inside the parent
+					for (auto initial_child : child->children)
+						initial_node->children.push_back(initial_child);
+				}
+			} else {
+				// AST_BLOCK in parent does not exist, place the contents
+				// of child initial node inside the parent
+				for (auto initial_child : child->children)
+					initial_node->children.push_back(initial_child);
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: Are this cases even possible? This AST_BLOCK should be created when creating AST_INITIAL, so I think we should check/assert that in ``child``/``parent`` node there is AST_BLOCK as a first children rather then this.

File: frontends/uhdm/UhdmAst.cc [hunk 10]
```diff
@@ -248,8 +248,55 @@ static void add_or_replace_child(AST::AstNode* parent, AST::AstNode* child) {
...
+				if (child_block_node_it != child->children.end()) {
+					// Place the contents of child initial node inside the parent initial
+					// But place the contents of child block node inside parent block
+					AST::AstNode* child_block_node = *child_block_node_it;
+					for (auto initial_child : child->children) {
+						if (initial_child == child_block_node) {
+							for (auto child_block_child : child_block_node->children)
+								block_node->children.push_back(child_block_child);
+						} else {
+							initial_node->children.push_back(initial_child);
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: I think we need to clone nodes here, when we are coping, as they can be deleted when node is replaced.

**Review by kamilrakoczy — CHANGES_REQUESTED** [review]
Some more comments to  changes in ``add_or_replace_child``.

File: frontends/uhdm/UhdmAst.cc (hunk 7 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: I didn't find any evidence in code that will prove that there can be other nodes than AST_BLOCK as AST_INITIAL children so I think that You are correct. I modified the code to check only if there are any children in AST_INITIAL and if the first child is AST_BLOCK. If those conditions aren't met we throw errors with `log_assert()`. This change greatly simplified the code, thank You for the idea!

File: frontends/uhdm/UhdmAst.cc (hunk 8 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Modified accordingly

File: frontends/uhdm/UhdmAst.cc (hunk 9 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Added checks and throwing errors when conditions aren't met

File: frontends/uhdm/UhdmAst.cc (hunk 10 above)
Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: Nodes are now cloned

**Comment by lpawelcz** [comment]
I'm now addressing failures in tests after recent changes

File: frontends/uhdm/UhdmAst.cc [hunk 11]
```diff
@@ -242,8 +248,44 @@ static void add_or_replace_child(AST::AstNode* parent, AST::AstNode* child) {
...
+
+			// Place the contents of child initial node inside the parent initial
+			// And place the contents of child block node inside parent block
+			for (auto initial_child : child->children) {
+				if (initial_child == child_block_node) {
+					for (auto child_block_child : child_block_node->children)
+						block_node->children.push_back(child_block_child->clone());
+				} else {
+					initial_node->children.push_back(initial_child->clone());
+				}
```
Comment (kamilrakoczy) [inline frontends/uhdm/UhdmAst.cc]: With recent changes, here we already know, that there is AST_BLOCK as first child of ``child`` node, so I think this if is unnecessary, and we can just always copy nodes.

If it is not the case, here we are comparing pointers, but we shouldn't do that. This can introduce 2 ``AST_BLOCK`` inside ``AST_INITIAL``.

Comment (lpawelcz) [reply frontends/uhdm/UhdmAst.cc]: I removed the if statement. Now we first clone the contents of child block into parent block and then proceed with copying other children nodes of the child node into parent initial node.

**Comment by lpawelcz** [comment]
Found a bug that most probably caused those all failures. Fixed it and hopefully all tests should pass now.

**Review by kamilrakoczy — APPROVED** [review]
LGTM
